[FLINK-19911] add read buffer for input stream

1996fanrui opened the pull request:
## What is the purpose of the change

Read checkpoint stream with buffer to speedup restore.

There will be many small IOs in the restore process,  frequent small io is not friendly to disk. So add buffer to improve IO performance.

## Brief change log

  - add config `state.backend.fs.read-buffer-size`
  - Design FSDataBufferedInputStream, FSDataInputStream with buffer
  - Use FSDataBufferedInputStream to wrap fsDataInputStream in HeapRestoreOperation#restore.

tips: 
The design of FSDataBufferedInputStream refers to FsCheckpointStateOutputStream and Java's BufferedInputStream.
Compared to Java's BufferedInputStream, the seek and getPos interfaces are mainly added.

## Verifying this change

This change added tests and can be verified as follows:

  - Added order read for FSDataBufferedInputStream
  - Added seek and random read for FSDataBufferedInputStream

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsStateBackend.java:
```diff
@@ -524,13 +524,17 @@ public CheckpointStorageAccess createCheckpointStorage(JobID jobId) throws IOExc
 		HeapPriorityQueueSetFactory priorityQueueSetFactory =
 			new HeapPriorityQueueSetFactory(keyGroupRange, numberOfKeyGroups, 128);
 
+		ReadableConfig config = env.getTaskManagerInfo().getConfiguration();
+		int fsReadBufferSize = config.get(CheckpointingOptions.FS_READ_BUFFER_SIZE);
```
Do not do this here, you should use the config passed to the `configure` method. Look at how the async flag is set.

1996fanrui replied:
Hi, @sjwiesman , thanks for your comments. I will do it.

1996fanrui replied:
done

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/memory/MemoryStateBackend.java:
```diff
@@ -331,13 +331,18 @@ public OperatorStateBackend createOperatorStateBackend(
 		TaskStateManager taskStateManager = env.getTaskStateManager();
 		HeapPriorityQueueSetFactory priorityQueueSetFactory =
 			new HeapPriorityQueueSetFactory(keyGroupRange, numberOfKeyGroups, 128);
+
+		ReadableConfig config = env.getTaskManagerInfo().getConfiguration();
+		int fsReadBufferSize = config.get(CheckpointingOptions.FS_READ_BUFFER_SIZE);
```
Same as above, please do this inside of `configure`.

1996fanrui replied:
done

sjwiesman commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/mock/MockStateBackend.java:
```diff
@@ -134,13 +136,18 @@ public CheckpointStreamFactory resolveCheckpointStorageLocation(long checkpointI
 		MetricGroup metricGroup,
 		@Nonnull Collection<KeyedStateHandle> stateHandles,
 		CloseableRegistry cancelStreamRegistry) {
+
+		Configuration configuration = env.getTaskManagerInfo().getConfiguration();
+		int fsReadBufferSize = configuration.getInteger(CheckpointingOptions.FS_READ_BUFFER_SIZE);
```
Same

1996fanrui replied:
done

sjwiesman commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -540,6 +540,10 @@ public CheckpointStorageAccess createCheckpointStorage(JobID jobId) throws IOExc
 
 		ExecutionConfig executionConfig = env.getExecutionConfig();
 		StreamCompressionDecorator keyGroupCompressionDecorator = getCompressionDecorator(executionConfig);
+
+		ReadableConfig config = env.getTaskManagerInfo().getConfiguration();
+		int fsReadBufferSize = config.get(CheckpointingOptions.FS_READ_BUFFER_SIZE);
```
Same

1996fanrui replied:
done

sjwiesman commented on flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java:
```diff
@@ -157,4 +157,14 @@
+	@Documentation.Section(Documentation.Sections.EXPERT_STATE_BACKENDS)
+	public static final ConfigOption<Integer> FS_READ_BUFFER_SIZE = ConfigOptions
+		.key("state.backend.fs.read-buffer-size")
+		.intType()
```
Shouldn't the type of this be `memory`? I know some of the others aren't but that's because they predate us having a memory configuration type.

1996fanrui replied:
This configuration only means reading a buffer size of the file system. Should the `memory configuration` contain some `memory size` related config?

Myasuka replied:
I am also in favor of `memory type` as the buffer is just an array of bytes allocated from JVM memory. The reason why `state.backend.fs.write-buffer-size` is integer type is that it implemented before Flink introduced `memory type` configuration.

1996fanrui replied:
done

sjwiesman reviewed (changes requested):
I've left a few comments. Please remove the merge commits from this branch and check if the test failure is relevant to this change.

1996fanrui commented:
@flinkbot  run azure

klion26 commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/filesystem/FSDataBufferedInputStreamTest.java:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	@Test
+	public void testSeek() throws Exception {
```
Do we need to add some bound test cases for `seek()` (such as seek out of bounding) here?

1996fanrui replied:
@klion26 , thanks for your suggest.
I have finished it. I hope you can help review again when you are free.

1996fanrui commented:
> I've left a few comments. Please remove the merge commits from this branch and check if the test failure is relevant to this change.

@sjwiesman Thanks for your review.
I have removed these merge commits and modified the initialization of config. 
Please help me to review again when you are free. Thanks.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,189 @@
+		this.inputStream = inputStream;
+
+		Preconditions.checkState(bufferSize > 0,
+			new IllegalArgumentException("bufferSize must > 0"));
```
We should pass error message instead of `IllegalArgumentException` here.

1996fanrui replied:
done

Myasuka commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -264,12 +270,36 @@ public RocksDBStateBackend(StateBackend checkpointStreamBackend) {
+	public RocksDBStateBackend(
+			StateBackend checkpointStreamBackend,
+			TernaryBoolean enableIncrementalCheckpointing,
+			int fsReadBufferSize) {
```
I don't think we should add a new constructor here. Putting `fsReadBufferSize` as one of the property in `FsStateBackend` is enough.

Myasuka commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -877,6 +913,20 @@ public long getWriteBatchSize() {
+	 *
+	 * @return The file system read buffer size, in bytes.
+	 */
+	public int getFsReadBufferSize() {
```
As you can see, there existed three duplicate `getFsReadBufferSize` in your PR, which could be an abstract base method for all state backends. 
If we think it one step further, we now has same interfaces for checkpoint write of state backend, e.g. `createCheckpointStorage`, however, we do not have same interfaces for checkpoint **read** of state backend. That's why you meet such embarrassment here. 
I think this deserve a discussion over FLINK-19465 and FLIP-142, @sjwiesman what do you think of this?

1996fanrui replied:
Thanks to @Myasuka 's perfect analysis, this is indeed the reason for my design.

Myasuka reviewed (changes requested):
Thanks for your contribution, please view my comments.

From this PR, we also see scenario to introduce basics interface for checkpoint reading, which should involve more wider discussion.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,189 @@
+/**
+ * FSDataInputStream with Buffer
+ */
+public class FSDataBufferedInputStream extends FSDataInputStream {
```
Thanks for your review.
Some code is copied from `java.io.BufferedInputStream`. Compared to `java.io.BufferedInputStream`, the seek and getPos interfaces are mainly added.
I will add some notes.

1996fanrui replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,189 @@
+
+	private static final int DEFAULT_BUFFER_SIZE = 8192;
+
+	protected volatile byte[] buf;
```
done

StephanEwen commented:
Sorry to be late to the game here, but could you share a bit more information on what the original setup was?
Specifically, what was your checkpoint storage system that offered such bad stream read performance? Was it HDFS? OSS? S3?

Looking at this change here, it seems very big with more options and many changed classes, for "just" introducing a buffer in a stream. That makes me skeptical that this is fixed in the right place.

Two other options to solve this:

(1) Input stream buffering is a property of the `CheckpointStorage`. It is created there, rather than in the state backends that have to wrap the stream. That way it works for all users of the `CheckpointStorage`, not just the state backends that happened to be adjusted to wrap the stream.

(2) Alternatively, we can make it a contract that all FileSystem implementations return well-buffered streams. Some already do this by default, wrapping them with another buffered stream adds just another layer and extra copying of bytes, costing performance. The ones that do not do that are easily adjusted.

At a first glance, I'd say let's go with option (2) if possible, otherwise option (1).
Hence the question: Which FS did you use that had such bad performing streams?

1996fanrui commented:
> Sorry to be late to the game here, but could you share a bit more information on what the original setup was?
> Specifically, what was your checkpoint storage system that offered such bad stream read performance? Was it HDFS? OSS? S3?
> 
> Looking at this change here, it seems very big with more options and many changed classes, for "just" introducing a buffer in a stream. That makes me skeptical that this is fixed in the right place.
> 
> Two other options to solve this:
> 
> (1) Input stream buffering is a property of the `CheckpointStorage`. It is created there, rather than in the state backends that have to wrap the stream. That way it works for all users of the `CheckpointStorage`, not just the state backends that happened to be adjusted to wrap the stream.
> 
> (2) Alternatively, we can make it a contract that all FileSystem implementations return well-buffered streams. Some already do this by default, wrapping them with another buffered stream adds just another layer and extra copying of bytes, costing performance. The ones that do not do that are easily adjusted.
> 
> At a first glance, I'd say let's go with option (2) if possible, otherwise option (1).
> Hence the question: Which FS did you use that had such bad performing streams?

Thanks for your comments.
I use hdfs, mainly because there are a lot of small IO during restore. A Flink job is provided in jira to reproduce the test results.

StephanEwen commented:
Are the HDFS input streams generally not buffered? Would it make sense to adjust the `HadoopDataInputStream` class to be buffered?

1996fanrui commented:
> Are the HDFS input streams generally not buffered? Would it make sense to adjust the `HadoopDataInputStream` class to be buffered?

Hi @StephanEwen , I am sorry to reply you so late.
In the past two days, I read the relevant code of hdfs reading data and did some performance analysis.
conclusion as below:
The hdfs input stream has buffer by default. The default buffer size is 64KB.
In theory, Use FSDataBufferedInputStream to wrap hdfsInputStream does not reduce the number of disk accesses. But the result shows: When buffer is added, restore time is reduced to one-third of the original.

So, I did some performance analysis.
From the point of view of CPU usage, When buffer is added cpu usage is greatly reduced.
Using hdfsInputStream directly, the CPU usage rate is 60~70%.
Use FSDataBufferedInputStream to wrap hdfsInputStream, the CPU usage rate is 20~25%.

Analyze why hdfsInputStream consumes CPU:
The hdfs client contains a lot of statistical information, and the method call stack is relatively deep, each method will consume a little performance. As shown in the picture below:
[flame graph remark link](https://drive.google.com/file/d/1zTwHdmSybAgyBGIIP71FLfMQ5DvK1eON/view?usp=sharing)

If you use FSDataBufferedInputStream to wrap hdfsInputStream, there will be a buffer outside the hdfs client, avoiding a very deep call stack and a lot of hdfs statistics.

Original flame graph link:
[restore with buffer](https://drive.google.com/file/d/1jxBNzh2iIsrX__wfFjvTWiC9LPCCDS3A/view?usp=sharing)
[restore without buffer](https://drive.google.com/file/d/1qDbqQC4bG34_ZsCrOpDaNouNQpvVXRWH/view?usp=sharing)

StephanEwen commented:
Thanks a lot for that deep diagnosis.

Given that result, would it make sense to buffer all Hadoop streams in Flink (also affecting S3, etc.), or just the DFS streams (HDFS)? Do all streams have such high statistics/metrics costs, or just the HDFS input stream?

In any case, it looks like we should add a `BufferedHadoopDataInputStream` that is a slight modification of the `HadoopDataInputStream` class with an internal buffer, and use that in the `HadoopFileSystem` when returning a stream. That way we should cover all cases.

1996fanrui commented:
> Thanks a lot for that deep diagnosis.
> 
> Given that result, would it make sense to buffer all Hadoop streams in Flink (also affecting S3, etc.), or just the DFS streams (HDFS)? Do all streams have such high statistics/metrics costs, or just the HDFS input stream?
> 
> In any case, it looks like we should add a `BufferedHadoopDataInputStream` that is a slight modification of the `HadoopDataInputStream` class with an internal buffer, and use that in the `HadoopFileSystem` when returning a stream. That way we should cover all cases.

It may be to make sense to buffer most of the streams in Flink.

First of all, it is beneficial to add buffer to `HadoopDataInputStream`. Secondly, there is no buffer in LocalDataInputStream, adding buffer can also improve IO performance. There will be some benefits for LocalRecovery.

For S3, I don't know much, and there is no S3 environment. I may not have a say.

1996fanrui commented:
I feel it can buffer all streams in Flink.

If `bufferSize> 0` then add buffer. If `bufferSize == 0`, the original InputStream is used. So if the buffer has a negative impact, the user can manually close the buffer. In theory, the buffer should not have a negative impact.

Please correct me if something I am missed.

1996fanrui commented:
@StephanEwen Do you have any ideas? Hope to get your guidance.

StephanEwen commented:
Yes, I think that makes sense. Adding a `fs.buffer` config option, maybe with a default value of 64k?

  - For local streams, it would be good as well, that is true.
  - For S3, we get that implicitly if we buffer the `HadoopDataInputStream`, because the S3 reads go through some Hadoop utilities and also wrap the stream with the `HadoopDataInputStream` in the end.

1996fanrui commented:
> Yes, I think that makes sense. Adding a `fs.buffer` config option, maybe with a default value of 64k?
> 
> * For local streams, it would be good as well, that is true.
> * For S3, we get that implicitly if we buffer the `HadoopDataInputStream`, because the S3 reads go through some Hadoop utilities and also wrap the stream with the `HadoopDataInputStream` in the end.

@StephanEwen Thanks for your reply.
After the above analysis, I think `FSDataBufferedInputStream` can wrap all InputStream.
Now there is a question: we do not have same interfaces for checkpoint read of state backend. How to pass `buffer size`?

There are three options:
1. Wait for the completion of [FLINK-19465](https://issues.apache.org/jira/browse/FLINK-19465), and then continue the current PR work
2. The current PR does not support modification of `buffer size`, and is set to 64KB uniformly.
3. Use three duplicate `getFsReadBufferSize` to complete the current PR, and then refactor in [FLINK-19465](https://issues.apache.org/jira/browse/FLINK-19465), or create a new issue to refactor.

what do you think of this? @Myasuka @sjwiesman

StephanEwen commented:
This should be totally independent from any checkpoint code, because it has nothing to do with checkpointing. This is purely file system business.

The changes would be in the `FileSystem` class and the `FileSystemFactory` picks up the configuration value in the `configure()` method.

1996fanrui commented:
> This should be totally independent from any checkpoint code, because it has nothing to do with checkpointing. This is purely file system business. I think we need to start a new implementation here, this PR is not in the right direction.
> 
> The changes would be in the `FileSystem` class and the `FileSystemFactory` picks up the configuration value in the `configure()` method.

Ok, I will read `FileSystem` class. I will do my best to do it.
If I have any questions, I will ask you again.

1996fanrui commented:
Hi @StephanEwen 
I've updated the PR based on your comments, Please help reviewing it when you have time.
Thanks.

AHeise commented:
Hi @1996fanrui , 

that's an interesting discovery and investigation that you did there!

I think the approach on the filesystem level is also much better than the previous way. Let's try to not change any public API (FileSystem) as this would slow down the progress.

I'd probably focus on Hadoop file systems entirely (for now). What I'd propose is the following:
- Use `HadoopFsFactory#configure` to extract the buffer size and pass it to the `ctor` of all filesystems created by the factory.
- Use that default buffer size in `HadoopFileSystem#open(Path)` to call `#open(Path, int)`.
- `HadoopFileSystem#open(Path, int)` should use the buffer size both in the call to Hadoop and to wrap it as you did as in the `BufferedFSInputStream`. I dug a bit into the Hadoop code and noticed that the cache is by default just 4kb. So even if we have cache on top of it with 64kb, we would still need to ask Hadoop several times. 

So, that means you are not adding any new methods, but just modify existing ones.

1996fanrui commented:
> Hi @1996fanrui ,
> 
> that's an interesting discovery and investigation that you did there!
> 
> I think the approach on the filesystem level is also much better than the previous way. Let's try to not change any public API (FileSystem) as this would slow down the progress.
> 
> I'd probably focus on Hadoop file systems entirely (for now). What I'd propose is the following:
> 
> * Use `HadoopFsFactory#configure` to extract the buffer size and pass it to the `ctor` of all filesystems created by the factory.
> * Use that default buffer size in `HadoopFileSystem#open(Path)` to call `#open(Path, int)`.
> * `HadoopFileSystem#open(Path, int)` should use the buffer size both in the call to Hadoop and to wrap it as you did as in the `BufferedFSInputStream`. I dug a bit into the Hadoop code and noticed that the cache is by default just 4kb. So even if we have cache on top of it with 64kb, we would still need to ask Hadoop several times.
> 
> So, that means you are not adding any new methods, but just modify existing ones.

Hi @AHeise , thanks for your suggestions.

In the process of modifying the code, I also felt that the public API was modified. There may be some unknown effects.

I have some questions now:  This feature can optimize the IO performance of `LocalFileSystem`, so can it be applied to `LocalFileSystem` and `HadoopFileSystem`?

AHeise commented:
> `LocalFileSystem`

Yes, it looks like `LocalFileSystem` should also be a good candidate. In particular `#open(Path, int)` currently ignores the buffer size...

I'm wondering if it makes more sense to create two PRs though. Even though the solution is pretty much the same, the underlying issues are different imho. The smaller a PR, the faster we can usually go. However, since they are so related, I wouldn't mind both issues being resolved in the same PR.

1996fanrui commented:
> > `LocalFileSystem`
> 
> Yes, it looks like `LocalFileSystem` should also be a good candidate. In particular `#open(Path, int)` currently ignores the buffer size...
> 
> I'm wondering if it makes more sense to create two PRs though. Even though the solution is pretty much the same, the underlying issues are different imho. The smaller a PR, the faster we can usually go. However, since they are so related, I wouldn't mind both issues being resolved in the same PR.

Hi @AHeise , thanks for your quick reply.  

I think it is okay to create two PRs.  The current PR developed `FsDataBuefferedInputStream` and applied it to `HadoopFileSysytem`.  

The next PR considers whether other FileSystem are profitable.

StephanEwen commented:
Thanks @AHeise for helping out here. Sorry @1996fanrui I am pretty overloaded the last days, so I could not follow up in a timely manner.

1996fanrui commented:
@StephanEwen Thank you for your very useful suggestions in the last few days, and I wish you a happy work.

1996fanrui commented:
> > > `LocalFileSystem`
> > 
> > 
> > Yes, it looks like `LocalFileSystem` should also be a good candidate. In particular `#open(Path, int)` currently ignores the buffer size...
> > I'm wondering if it makes more sense to create two PRs though. Even though the solution is pretty much the same, the underlying issues are different imho. The smaller a PR, the faster we can usually go. However, since they are so related, I wouldn't mind both issues being resolved in the same PR.
> 
> Hi @AHeise , thanks for your quick reply.
> 
> I think it is okay to create two PRs. The current PR developed `FsDataBuefferedInputStream` and applied it to `HadoopFileSysytem`.
> 
> The next PR considers whether other FileSystem are profitable.

Hi @AHeise ,

Do you think this plan is OK? If ok, I will develop the code of the current PR when I have time.

AHeise commented:
> > > > `LocalFileSystem`
> > > 
> > > 
> > > Yes, it looks like `LocalFileSystem` should also be a good candidate. In particular `#open(Path, int)` currently ignores the buffer size...
> > > I'm wondering if it makes more sense to create two PRs though. Even though the solution is pretty much the same, the underlying issues are different imho. The smaller a PR, the faster we can usually go. However, since they are so related, I wouldn't mind both issues being resolved in the same PR.
> > 
> > 
> > Hi @AHeise , thanks for your quick reply.
> > I think it is okay to create two PRs. The current PR developed `FsDataBuefferedInputStream` and applied it to `HadoopFileSysytem`.
> > The next PR considers whether other FileSystem are profitable.
> 
> Hi @AHeise ,
> 
> Do you think this plan is OK? If ok, I will develop the code of the current PR when I have time.

Yes @1996fanrui , please go ahead.

Myasuka commented:
As we use `OSSFileSystemFactory` for our production usage and it also wrapped as `HadoopFileSystem`, I think it's okay to introduce buffer in `HadoopFileSystem` and most DFS could benefit.

1996fanrui commented:
Hi @AHeise  @Myasuka ,Thanks for your reply. I will do it.

1996fanrui commented:
> Hi @1996fanrui ,
> 
> that's an interesting discovery and investigation that you did there!
> 
> I think the approach on the filesystem level is also much better than the previous way. Let's try to not change any public API (FileSystem) as this would slow down the progress.
> 
> I'd probably focus on Hadoop file systems entirely (for now). What I'd propose is the following:
> 
> * Use `HadoopFsFactory#configure` to extract the buffer size and pass it to the `ctor` of all filesystems created by the factory.
> * Use that default buffer size in `HadoopFileSystem#open(Path)` to call `#open(Path, int)`.
> * `HadoopFileSystem#open(Path, int)` should use the buffer size both in the call to Hadoop and to wrap it as you did as in the `BufferedFSInputStream`. I dug a bit into the Hadoop code and noticed that the cache is by default just 4kb. So even if we have cache on top of it with 64kb, we would still need to ask Hadoop several times.
> 
> So, that means you are not adding any new methods, but just modify existing ones.

Hi @Myasuka @AHeise ,

Based on the above design, there are some questions that I hope to be answered:

- `buffer size` is passed to HadoopFileSystem through Constructor. It means that HadoopFileSystem needs to add a new Constructor: `HadoopFileSystem(FileSystem, bufferSize)`. HadoopFsFactory will call the new constructor.

- The old constructor `HadoopFileSystem(FileSystem)` is still called by other FsFactory. So other FsFactory will not be able to improve performance. For example: OSSFileSystemFactory.

Question: Do other FsFactory related to HadoopFileSystem need to improve performance? If needed, are there other better designs. (Can there be a way not to modify many FsFactory?)

Thanks.

AHeise commented:
> Based on the above design, there are some questions that I hope to be answered:
> 
>     * `buffer size` is passed to HadoopFileSystem through Constructor. It means that HadoopFileSystem needs to add a new Constructor: `HadoopFileSystem(FileSystem, bufferSize)`. HadoopFsFactory will call the new constructor.
> 
>     * The old constructor `HadoopFileSystem(FileSystem)` is still called by other FsFactory. So other FsFactory will not be able to improve performance. For example: OSSFileSystemFactory.
> 
> 
> Question: Do other FsFactory related to HadoopFileSystem need to improve performance? If needed, are there other better designs. (Can there be a way not to modify many FsFactory?)
> 
> Thanks.

Similar to your initial design, I would make `bufferSize = 0` => no buffering. Then all other filesystem factories could use the changed constructor with `HadoopFileSystem(FileSystem, NO_BUFFER)` (`NO_BUFFER = 0`).

Since this PR is very specific to Hadoop, I'd probably set the buffer only through `HadoopFsFactory`. Your finding would not necessarily also transfer to other implementations of the `hadoop.FileSystem` (i.e., they don't have statistics). We could evaluate that in future PRs.

1996fanrui commented:
Hi @AHeise ,
I've updated the PR, Please help reviewing it when you have time.
Thanks.

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+
+	private boolean closed;
+
+	public FSDataBufferedInputStream(FSDataInputStream inputStream) {
+		this(inputStream, DEFAULT_BUFFER_SIZE);
+	}
+
```
This constructor is only used in tests, I'd remove it here and add the constant to the test.

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+
+	private static final int DEFAULT_BUFFER_SIZE = 8192;
+
+	protected byte[] buf;
```
Couldn't that be private?

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	public FSDataBufferedInputStream(
+		FSDataInputStream inputStream,
+		int bufferSize) {
```
nit: That should fit on one line.

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+	public FSDataBufferedInputStream(
+		FSDataInputStream inputStream,
+		int bufferSize) {
+		this.inputStream = inputStream;
```
this.inputStream = Preconditions.checkNotNull(inputStream);

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+
+	@Override
+	public void close() throws IOException {
+		if (closed) {
+			return;
+		}
+		closed = true;
+		if (inputStream != null) {
+			inputStream.close();
+		}
+	}
```
I'd set `inputStream = null` after closing. Then you don't need `closed` at all - you could simply check if `inputStream != null`.

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	@Override
+	public void seek(long desired) throws IOException {
```
All operations should check if the stream has already been closed. I'd add `checkClosed();` as the first line to pretty much all methods.
See also my comment below - you don't need `closed`. `checkClosed could throw an IllegalStateException whenever, `inputStream == null`.

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+
+	private void fill() throws IOException {
+		Preconditions.checkState(pos >= count);
+		count = inputStream.read(buf);
```
count can be `-1` here. Please double-check if it's used correctly in all places. `getPos` seems to not work for example. `seek` also looks suspicious.

AHeise commented on flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java:
```diff
@@ -0,0 +1,196 @@
+		long avail = count - pos;
+
+		if (avail <= 0) {
+			// Fill in buffer to save bytes for reset
```
It feels as if we could improve `skip` by using `inputStream.skip`. Currently, `skip` is pretty much just `read` with discarding bytes. At least, S3InputStream and DFSInputStream have special implementations for skip.

AHeise commented on flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileSystem.java:
```diff
@@ -51,7 +55,22 @@
+	 * @param hadoopFileSystem The Hadoop FileSystem that will be used under the hood.
+	 * @param readBufferSize The size of the buffer to be used.
+	 */
+	public HadoopFileSystem(org.apache.hadoop.fs.FileSystem hadoopFileSystem,
+							int readBufferSize) {
```
nit: formatting (either bring on one line, or chop down both parameters with double indent.

AHeise commented on flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileSystem.java:
```diff
@@ -116,17 +135,19 @@ public FileStatus getFileStatus(final Path f) throws IOException {
 		final org.apache.hadoop.fs.FSDataInputStream fdis = this.fs.open(path, bufferSize);
-		return new HadoopDataInputStream(fdis);
+		HadoopDataInputStream inputStream = new HadoopDataInputStream(fdis);
+		if (readBufferSize <= 0) {
```
`readBufferSize == 0` is clearer -- it cannot be negative because of your ctor check. Also I'd add a constant for `0`. (`NO_BUFFER`, `NO_BUFFERING`, ...)

AHeise commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFsFactoryTest.java:
```diff
@@ -61,4 +72,61 @@ public void testCreateHadoopFsWithMissingAuthority() throws Exception {
+		final URI uri = URI.create("hdfs://localhost:12345/");
+		final Path path = new Path(uri.getPath());
+
+		org.apache.hadoop.fs.FileSystem hadoopFs = mock(org.apache.hadoop.fs.FileSystem.class);
+		org.apache.hadoop.fs.FSDataInputStream hadoopInputStream =
+			mock(org.apache.hadoop.fs.FSDataInputStream.class);
+		when(hadoopFs.open(isA(org.apache.hadoop.fs.Path.class), anyInt()))
+			.thenReturn(hadoopInputStream);
+
```
We actually want to move away from mocking with mockito as far as possible.
I'm wondering what's the benefit over using `file://` to some random (or temp) file?

AHeise commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFsFactoryTest.java:
```diff
@@ -61,4 +72,61 @@ public void testCreateHadoopFsWithMissingAuthority() throws Exception {
+		fsFactory.configure(configuration);
+
+		FileSystem fileSystem = fsFactory.create(uri);
+		mockHadoopFsOpen(fileSystem, hadoopFs);
```
I hope that we don't need that as well...

AHeise commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFsFactoryTest.java:
```diff
@@ -61,4 +72,61 @@ public void testCreateHadoopFsWithMissingAuthority() throws Exception {
 	}
+
+	@Test
+	public void testReadBufferSize() throws Exception {
```
I'd probably divide the test into different cases: One without setting property, one with setting it to 0, one with setting it to >0. The additional lines should make the tests much easier to read.

AHeise reviewed (commented):
Thank you again for your great contribution! I feel that we are almost there. Three things that I found:
- All methods of `FSDataBufferedInputStream` would kinda work on a closed stream, but that's unexpected.
- `FSDataBufferedInputStream.skip` should be improved - not sure if it's used somewhere, but you'll never know. It could mean a whole lot of difference in performance and costs when using columnar formats on S3 or similar.
- One tests uses mockito. We try to get rid of it entirely as it makes maintenance more complicated. If you have trouble getting the test without mockito running, I can help.

1996fanrui commented:
@AHeise 
Thanks for your review. I am very busy at work recently, and may address your comment later.

StephanEwen commented:
The PR looks to copy code from the Java SDK, which is Oracle licensed proprietary code, so this cannot be merged in any way.
@AHeise I think that is the reason for the unexpected formatting / visibility.

The PR also introduces again double wrapping: Hadoop's Stream -> `HadoopDataInputStream` -> `FSDataBufferedInputStream`

I would suggest to close this PR and go back to first defining what actually needs to be done, and how to approach this (not by copying copyrighted code!). Writing code here is the least important thing (introducing a buffer is not a complex problem) and it looks the PR takes a lot of iterations because it is always putting code in place before understanding how the problem should be approached.

1996fanrui commented:
@StephanEwen 
Sorry, because I ignored the copyright. I will close this PR.
